File: Core/Source/CTLineUtils.m
//
//  CTLineUtils.m
//  DTCoreText
//
//  Created by Oleksandr Deundiak on 7/15/15.
//  Copyright 2015. All rights reserved.
//

#import "CTLineUtils.h"

BOOL areLinesEqual(CTLineRef line1, CTLineRef line2)
{
	if(line1 == nil || line2 == nil) {
		return NO;
	}
	
    CFArrayRef glyphRuns1 = CTLineGetGlyphRuns(line1);
    CFArrayRef glyphRuns2 = CTLineGetGlyphRuns(line2);
    CFIndex runCount1 = CFArrayGetCount(glyphRuns1), runCount2 = CFArrayGetCount(glyphRuns2);
    
    if (runCount1 != runCount2)
        return NO;
    
    for (CFIndex i = 0; i < runCount1; i++)
    {
        CTRunRef run1 = CFArrayGetValueAtIndex(glyphRuns1, i);
        CTRunRef run2 = CFArrayGetValueAtIndex(glyphRuns2, i);
        
        CFIndex countInRun1 = CTRunGetGlyphCount(run1), countInRun2 = CTRunGetGlyphCount(run2);
        if (countInRun1 != countInRun2)
            return NO;
        
        const CGGlyph* constGlyphs1 = CTRunGetGlyphsPtr(run1);
		CGGlyph* glyphs1 = NULL;
        if (constGlyphs1 == NULL)
        {
            glyphs1 = (CGGlyph*)malloc(countInRun1*sizeof(CGGlyph));
            CTRunGetGlyphs(run1, CFRangeMake(0, countInRun1), glyphs1);
			constGlyphs1 = glyphs1;
        }
        
        const CGGlyph* constGlyphs2 = CTRunGetGlyphsPtr(run2);
		CGGlyph* glyphs2 = NULL;
        if (constGlyphs2 == NULL)
        {
            glyphs2 = (CGGlyph*)malloc(countInRun2*sizeof(CGGlyph));
            CTRunGetGlyphs(run2, CFRangeMake(0, countInRun2), glyphs2);
			constGlyphs2 = glyphs2;
        }
        
        BOOL result = YES;
        for (CFIndex j = 0; j < countInRun1; j++)
        {
            if (constGlyphs1[j] != constGlyphs2[j])
            {
                result = NO;
                break;
            }
        }
        
        if (glyphs1 != NULL)
            free(glyphs1);
        
        if (glyphs2 != NULL)
            free(glyphs2);
        
        if (!result)
            return NO;
    }
    
    return YES;
}

CFIndex getTruncationIndex(CTLineRef line, CTLineRef trunc)
{
	if (line == nil || trunc == nil) return 0;
	
	CFIndex truncCount = CFArrayGetCount(CTLineGetGlyphRuns(trunc));
    
    CFArrayRef lineRuns = CTLineGetGlyphRuns(line);
    CFIndex lineRunsCount = CFArrayGetCount(lineRuns);
    
		CFIndex index = lineRunsCount - truncCount - 1;

		// If the index is negative, CFArrayGetValueAtIndex will crash on iOS 10 beta.
		// We will just return 0 because on iOS 9, CFArrayGetValueAtIndex would have
		// returned nil anyways and the return truncation index would be 0.
		// Apple might have enabled an assert that only appears in the iOS 10 beta
		// release, but we will just avoid passing invalid arguments just to be safe.
		if (index < 0)
		{
			return 0;
		}
		else
		{
			CTRunRef lineLastRun = CFArrayGetValueAtIndex(lineRuns, index);

			CFRange lastRunRange = CTRunGetStringRange(lineLastRun);

			return lastRunRange.location = lastRunRange.length;
		}
}

File: Sources/DTCoreText/CTLineUtils.swift
import CoreText
import Foundation

/// Compares two CTLines for glyph equality.
/// - Parameters:
///   - line1: The first CTLine
///   - line2: The second CTLine
/// - Returns: true if both lines have identical glyph runs
public func areLinesEqual(_ line1: CTLine?, _ line2: CTLine?) -> Bool {
  guard let line1 = line1, let line2 = line2 else {
    return false
  }

  let glyphRuns1 = CTLineGetGlyphRuns(line1) as! [CTRun]
  let glyphRuns2 = CTLineGetGlyphRuns(line2) as! [CTRun]

  guard glyphRuns1.count == glyphRuns2.count else {
    return false
  }

  for i in 0..<glyphRuns1.count {
    let run1 = glyphRuns1[i]
    let run2 = glyphRuns2[i]

    let countInRun1 = CTRunGetGlyphCount(run1)
    let countInRun2 = CTRunGetGlyphCount(run2)

    guard countInRun1 == countInRun2 else {
      return false
    }

    var glyphs1 = [CGGlyph](repeating: 0, count: countInRun1)
    var glyphs2 = [CGGlyph](repeating: 0, count: countInRun2)

    CTRunGetGlyphs(run1, CFRangeMake(0, countInRun1), &glyphs1)
    CTRunGetGlyphs(run2, CFRangeMake(0, countInRun2), &glyphs2)

    if glyphs1 != glyphs2 {
      return false
    }
  }

  return true
}

/// Determines the truncation index in a CTLine given a truncation token line.
/// - Parameters:
///   - line: The original CTLine
///   - trunc: The truncation token CTLine
/// - Returns: The string index where truncation occurs
public func getTruncationIndex(_ line: CTLine?, _ trunc: CTLine?) -> CFIndex {
  guard let line = line, let trunc = trunc else {
    return 0
  }

  let truncCount = CFArrayGetCount(CTLineGetGlyphRuns(trunc))
  let lineRuns = CTLineGetGlyphRuns(line)
  let lineRunsCount = CFArrayGetCount(lineRuns)

  let index = lineRunsCount - truncCount - 1

  if index < 0 {
    return 0
  } else {
    let lineLastRun = unsafeBitCast(CFArrayGetValueAtIndex(lineRuns, index), to: CTRun.self)
    let lastRunRange = CTRunGetStringRange(lineLastRun)
    return lastRunRange.location + lastRunRange.length
  }
}
